kernel/common/enumerable.rb

```diff
@@ -408,11 +419,19 @@ def drop_while
   end
 
   def each_cons(num)
-    return to_enum(:each_cons, num) unless block_given?
-
     n = Rubinius::Type.coerce_to_collection_index num
     raise ArgumentError, "invalid size: #{n}" if n <= 0
 
+    return to_enum(:each_cons, num) {
+      if (enum_size = enumerator_size).nil? 
```

Comment:
Hm, that's a good point actually. I ran a quick benchmark (https://gist.github.com/YorickPeterse/da524a182febcd130436) which outputs the following:

```
Calculating -------------------------------------
          #each_cons    26.586k i/100ms
 #each_cons_no_cache    36.156k i/100ms
-------------------------------------------------
          #each_cons    480.690k (± 6.2%) i/s -      2.393M
 #each_cons_no_cache    448.608k (± 0.6%) i/s -      2.242M

Comparison:
          #each_cons:   480690.2 i/s
 #each_cons_no_cache:   448608.5 i/s - 1.07x slower
```

This shows that it indeed is a little bit faster when using a variable assignment, although not that much. Either way, the rationale is now clear to me. Having said that, I do think that perhaps moving the assignment out of the if statement makes it a bit easier to grok (but this could just be me):

```ruby
enum_size = enumerator_size

if enum_size.nil?

elsif ...

end
```